client/router-hooks.ts

```diff
@@ -0,0 +1,38 @@
+import { useLocation } from 'wouter'
+
+/**
+ * A hook that reads a search param with a given name from `window.location`, and allows changing
+ * it. Changing the search param will re-render the component.
+ *
+ * If there are multiple params with the same name (which is apparently supported by the standard),
+ * we return the first one.
+ *
+ * @example
+ *
+ * export const UserList = () => {
+ *   // /users?page=3
+ *   const [page, setPage] = useLocationSearchParam('page')
+ *   console.log(page) // '3'
+ *   return <button onClick={() => setPage('4')}>Go to page 4</button>
+ * }
+ */
+export const useLocationSearchParam = (
+  name: string,
+): [value: string, setValue: (value: string) => void] => {
+  const [location, setLocation] = useLocation()
+  const searchParams = new URLSearchParams(window.location.search)
```

Comment:
imo this should be memoized. And if you're going to use `URLSearchParams` you may as well just use `URL` on `location` instead (which then avoids accessing this global value randomly): https://developer.mozilla.org/en-US/docs/Web/API/URL/searchParams

Reply:
Hmm, I actually tried doing that initially (and now again after completely forgetting about it :p), but `wouter` doesn't return the query parameters in their `useLocation` hook. There's an open issue about it: https://github.com/molefrog/wouter/issues/58

Reply:
`useLocation` just gives you a string. I wasn't suggesting using it directly, I was telling you to use it to construct a `URL`, which will parse the string and give you parameters. I think you may have misread what I wrote, this is all I was suggesting:

```ts
const [location, setLocation] = useLocation()
const searchParams = useMemo(() => {
  const url = new URL(location)
  return location.searchParams
}, [location])
```

Reply:
Ah, I clicked through to the issue and I see what you mean now. That is extremely unexpected to me and I think we should probably write a hook that handles this properly, at the very least. I think there are other parts of the codebase that have bugs now because they expect query string changes to re-render

Reply:
As far as I can see in their source code, they do take query string into account when deciding when to re-render, so we should be safe there. They just don't return the query string from the `useLocation` hook for whatever reason.

We could provide our own hook that will be used for `useLocation`, but we'd basically have to reimplement everything they already did just to add `search` variable in the return array here: https://github.com/molefrog/wouter/blob/master/use-location.js#L60

Unless I'm missing something and there's a deeper issue here.

Reply:
Their implementation calls `dispatchEvent`, thus you can just listen on `window` for these events and not reimplement anything. There are examples in the linked issue of exactly how to do this